sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala

```diff
@@ -229,11 +231,15 @@ private[parquet] object ParquetTypesConverter extends Logging {
     case LongType => Some(ParquetTypeInfo(ParquetPrimitiveTypeName.INT64))
     case TimestampType => Some(ParquetTypeInfo(ParquetPrimitiveTypeName.INT96))
     case DecimalType.Fixed(precision, scale) if precision <= 18 =>
-      // TODO: for now, our writer only supports decimals that fit in a Long
       Some(ParquetTypeInfo(ParquetPrimitiveTypeName.FIXED_LEN_BYTE_ARRAY,
         Some(ParquetOriginalType.DECIMAL),
         Some(new DecimalMetadata(precision, scale)),
         Some(BYTES_FOR_PRECISION(precision))))
+    case DecimalType.Fixed(precision, scale) =>
+      Some(ParquetTypeInfo(ParquetPrimitiveTypeName.BINARY,
```

Comment:
Under the assumption that all values will use the full length, yes.

But at some point the overhead of the length is low compared to the overhead if someone specifies just the upper bound of values.
I have to check if it really uses 4 bytes for BINARY. I'd then raise the threshold to ~40 bytes length. (meaning <=10% worst case overhead before compression)

It won't simplify the decoding/writing though, because the <=18 case  is used for long decoding.